refextract/references/pdf.py

```diff
@@ -48,40 +47,119 @@ def extract_texkeys_from_pdf(pdf_file):
+                def uri_sortfunc(uri_obj):
+                    page, _, ypos, xpos = uri_obj[1]
+                    return (page, ypos, xpos)
+
+            refs = [(dest[0], _destination_position(pdf, dest[1])) for dest in refs]
+            refs.sort(key=uri_sortfunc)
+            urls = [
+                (uri["/A"]["/URI"], _uri_destination_position(pdf, uri)) for uri in urls
+            ]
+            urls.sort(key=uri_sortfunc)
+            texkey_url_list = []
+            for nb, ref in enumerate(refs):
+                current_texkey_urls_dict = {}
+                current_texkey_urls_dict["texkey"] = re_reference_in_dest.match(
+                    ref[0]
+                ).group(1)
+                if nb < len(refs) - 1:
+                    next_reference_data = refs[nb + 1]
+                    matched_urls_for_reference, urls = _match_url_with_references(
+                        urls, ref, next_reference_data
+                    )
+                else:
+                    matched_urls_for_reference = _match_last_reference(urls, ref)
+                if matched_urls_for_reference:
+                    current_texkey_urls_dict["urls"] = matched_urls_for_reference
+                texkey_url_list.append(current_texkey_urls_dict)
+
+            return texkey_url_list
         except Exception:
             LOGGER.debug(u"PDF: Impossible to determine layout, no TeXkeys returned")
             return []
 
 
+def _match_url_with_references(urls_to_match, reference, next_reference):
+    ref_page_number, ref_column, ref_y, _ = reference[1]
+    next_ref_page_number, next_ref_col, next_ref_y, _ = next_reference[1]
+    urls_for_reference = set()
+    url_index = -1
+    for url in urls_to_match:
+        url_index += 1
+        url_page_number, _, url_y, _ = url[1]
+        # url between texkeys
+        if (
+            ref_page_number == url_page_number or ref_page_number + 1 == url_page_number
+        ) and (ref_y < url_y < next_ref_y):
+            urls_for_reference.add(url[0])
+        # last reference in a page
+        elif (
+            (ref_page_number == url_page_number) and
+            (next_ref_page_number > url_page_number) and
+            (ref_y < url_y)
+        ):
+            urls_for_reference.add(url[0])
+        # same page, next column
+        elif (
+            (ref_page_number == url_page_number) and
+            (next_ref_page_number == url_page_number) and
+            (ref_y < url_y) and
+            (next_ref_col > ref_column)
+        ):
+            urls_for_reference.add(url[0])
+        # case when url doesn't correspond to any reference
+        elif ref_page_number > url_page_number:
+            continue
+        # case when we should start processing a new reference
+        elif url_y > next_ref_y:
+            urls_to_match = urls_to_match[url_index:]
+            break
+
+    return urls_for_reference, urls_to_match
+
+
+def _match_last_reference(urls_to_match, reference):
```

Comment:
do you need a separate function here? couldn't you handle this by passing `None` to `_match_url_with_references` and handling that case there?